File: day-7/python/main.py
import sys


class File(object):
    def __init__(self, name, size):
        self.name = name
        self.size = size

class Dir(object):
    def __init__(self, name, parent):
        self.name = name
        self.parent = parent
        self.dirs = []
        self.files = []
        self._size = None

    def filesize(self):
        if self._size is None:
            s = 0
            for f in self.files:
                s += f.size
            for d in self.dirs:
                s += d.filesize()
            self._size = s
        return self._size


def parse_change_dir(s, cwd, root):
    if s[2] == '..':
        assert cwd is not None
        cwd.parent.dirs.append(cwd)
        cwd = cwd.parent
    else:
        nd = Dir(s[2], cwd)
        if cwd is None:
            assert s[2] == '/'
            root = nd
        cwd = nd
    return cwd, root


def parse_files(fd, cwd):
    while True:
        line = fd.readline()
        if not line:
            cwd.parent.dirs.append(cwd)
            break
        if line.startswith('$'):
            break
        s = line.split()
        if s[0] == 'dir':
            pass
        else:
            cwd.files.append(File(s[1], int(s[0])))
    return cwd, line


def parse_input(filename):
    with open(filename) as fd:
        cwd = None
        line = fd.readline()
        root = None
        while line:
            assert(line.startswith('$'))
            s = line.split()
            if s[1] == 'cd':
                cwd, root = parse_change_dir(s, cwd, root)
                line = fd.readline()
            else:
                assert s[1] == 'ls'
                cwd, line = parse_files(fd, cwd)
    return root


def problem_1(root):
    s = 0
    st = [root]
    while True:
        try:
            d = st.pop()
        except:
            break
        if d.filesize() < 100000:
            s += d.filesize()
        for t in d.dirs:
            st.append(t)
    return s


def problem_2(root):
    total = 70000000
    needed = 30000000
    tofree = needed - (total - root.filesize())
    s = sys.maxsize
    st = [root]
    while True:
        try:
            d = st.pop()
        except:
            break
        if d.filesize() > tofree:
            s = min(s, d.filesize())
        for t in d.dirs:
            st.append(t)
    return s


def parse_args():
    if len(sys.argv) != 2:
        sys.exit('I need a file and just a file')
    return sys.argv[1]


def main():
    filename = parse_args()
    root = parse_input(filename)
    print('Day 7, problem 1: ', problem_1(root))
    print('Day 7, problem 2: ', problem_2(root))


if __name__ == '__main__':
    main()

File: day-7/rust/src/main.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::process;
use std::collections::HashMap;

fn parse_args() -> String {
    let mut args = env::args();
    if args.len() != 2 {
        eprintln!("Invalid arguments: I need the input file and nothing else");
        process::exit(1);
    }
    args.nth(1).unwrap()
}

fn create_reader(filename: &String) -> impl Iterator<Item = String> {
    let file = File::open(&filename).unwrap_or_else(|_| {
        eprintln!("Cannot open file: {}", &filename);
        process::exit(1);
    });
    io::BufReader::new(file)
        .lines()
        .map(|l| l.unwrap())
}

fn change_dir(target: &str, dirs: &mut Vec::<String>) {
    if target == ".." {
        assert!(!dirs.is_empty());
        dirs.pop().unwrap();
    } else {
        let mut name = target.to_string();
        if dirs.len() > 1 {
            name = dirs[1..].join("/") + &name
        }
        dirs.push(name);
    }
}

fn sum_file_sizes(reader: &mut impl Iterator<Item = String>)
    -> (u64, Option<String>) {
    let mut file_sum: u64 = 0;
    let mut line = reader.next();
    while line.is_some() {
        let s = line.as_ref().unwrap();
        if s.starts_with("$") {
            break;
        }
        let mut p = s.split_whitespace();
        let a = p.next().unwrap();
        if a != "dir" {
            file_sum +=  a.parse::<u64>().unwrap();
        }
        line = reader.next();
    }
    (file_sum, line)
}

fn find_dir_sizes(filename: &String) -> HashMap::<String, u64> {
    let mut reader = create_reader(filename);
    let mut dir_sizes = HashMap::<String, u64>::new();
    let mut dirs = Vec::<String>::new();
    let mut line = reader.next();
    while line.is_some() {
        let s = line.as_ref().unwrap();
        assert!(&s.starts_with("$"));
        let mut p = s.split_whitespace();
        let cmd = p.nth(1).unwrap();
        if cmd == "cd" {
            let nd = p.next().unwrap();
            change_dir(&nd, &mut dirs);
            line = reader.next();
        } else {
            let (file_sum, nline) = sum_file_sizes(&mut reader);
            line = nline;
            // ls
            for d in &dirs {
                if let Some(x) = dir_sizes.get_mut(d) {
                    *x += file_sum;
                } else {
                    dir_sizes.insert(d.clone(), file_sum);
                }
            }
        }
    }
    dir_sizes
}

fn problem_1(dir_sizes: &HashMap::<String, u64>) -> u64 {
    dir_sizes.iter().map(|x| x.1).filter(|x| *x < &100000).sum()
}

fn problem_2(dir_sizes: &HashMap::<String, u64>) -> u64 {
    const TOTAL: u64 = 70000000;
    const NEEDED: u64 = 30000000;
    let tofree = NEEDED - (TOTAL - dir_sizes.get("/").unwrap());
    *dir_sizes.iter().map(|x| x.1).filter(|x| *x > &tofree).min().unwrap()
}

fn main() {
    let filename = parse_args();
    let dir_sizes = find_dir_sizes(&filename);

    println!("Day 7, problem 1: {}", problem_1(&dir_sizes));
    println!("Day 7, problem 2: {}", problem_2(&dir_sizes));
}
